src/main/java/net/neoforged/neoforge/capabilities/CapabilityListenerHolder.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright (c) NeoForged and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.neoforged.neoforge.capabilities;
+
+import it.unimi.dsi.fastutil.longs.Long2ReferenceMap;
+import it.unimi.dsi.fastutil.longs.Long2ReferenceOpenHashMap;
+import it.unimi.dsi.fastutil.objects.ReferenceOpenHashSet;
+import java.lang.ref.ReferenceQueue;
+import java.lang.ref.WeakReference;
+import java.util.Set;
+import net.minecraft.core.BlockPos;
+import net.minecraft.world.level.ChunkPos;
+import org.jetbrains.annotations.ApiStatus;
+
+// TODO: invalidate when levels are unloaded... maybe even when the server is stopped?
+@ApiStatus.Internal
+public class CapabilityListenerHolder {
+    private final Long2ReferenceMap<Long2ReferenceMap<Set<ListenerReference>>> byChunk = new Long2ReferenceOpenHashMap<>();
+    private final ReferenceQueue<ICapabilityInvalidationListener> queue = new ReferenceQueue<>();
+
+    public void addListener(BlockPos pos, ICapabilityInvalidationListener listener) {
+        pos = pos.immutable();
+        var chunkHolder = byChunk.computeIfAbsent(ChunkPos.asLong(pos), l -> new Long2ReferenceOpenHashMap<>());
+        chunkHolder.computeIfAbsent(pos.asLong(), l -> new ReferenceOpenHashSet<>()).add(new ListenerReference(queue, pos, listener));
+    }
+
+    public void invalidatePos(BlockPos pos) {
+        var chunkHolder = byChunk.get(ChunkPos.asLong(pos));
+        if (chunkHolder != null) {
+            var caches = chunkHolder.get(pos.asLong());
+            if (caches != null)
+                invalidateList(caches);
+        }
+    }
+
+    public void invalidateChunk(ChunkPos chunkPos) {
+        var chunkHolder = byChunk.get(chunkPos.toLong());
+        if (chunkHolder != null) {
+            for (var caches : chunkHolder.values())
+                invalidateList(caches);
+        }
+    }
+
+    private void invalidateList(Set<ListenerReference> caches) {
+        caches.removeIf(ref -> {
+            var listener = ref.get();
+            return listener == null || !listener.onInvalidate();
+        });
+    }
+
+    public void clean() {
+        while (true) {
+            ListenerReference ref = (ListenerReference) queue.poll();
+            if (ref == null)
+                return;
+
+            var chunkHolder = byChunk.get(ChunkPos.asLong(ref.pos));
+            if (chunkHolder == null)
+                continue;
+
+            var set = chunkHolder.get(ref.pos.asLong());
+            set.remove(ref);
+
+            if (set.isEmpty())
+                chunkHolder.remove(ref.pos.asLong());
+            if (chunkHolder.isEmpty())
+                byChunk.remove(ChunkPos.asLong(ref.pos));
```

Comment:
If we actually make use of the remove call's return result we can short circuit a bit of this logic:
```java
if (set.remove(ref) && set.isEmpty()) {
    chunkHolder.remove(ref.pos.asLong());
    if (chunkHolder.isEmpty())
        byChunk.remove(ChunkPos.asLong(ref.pos));
}
```